eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs

```diff
@@ -233,24 +233,32 @@ validateCollateralContainsNonADA ::
   Map.Map (TxIn (EraCrypto era)) (TxOut era) ->
   Test (AlonzoUtxoPredFailure era)
 validateCollateralContainsNonADA txBody utxoCollateral =
-  -- When we do not have any non-ada TxOuts we can short-circuit the more
-  -- expensive validation, which has to compute the full balance on the Value,
-  -- not just the Coin.
-  if areAllAdaOnly utxoCollateral && areAllAdaOnly (txBody ^. collateralReturnTxBodyL)
-    then pure ()
-    else failureUnless (Val.isAdaOnly bal) $
-      case txBody ^. collateralReturnTxBodyL of
-        SJust retTxOut
-          | not (Val.isAdaOnly colbal) ->
-              CollateralContainsNonADA (retTxOut ^. valueTxOutL)
-        _ -> CollateralContainsNonADA colbal
+  -- When we do not have any non-ada TxOuts we can short-circuit the more expensive
+  -- validation, which has to compute the full balance on the Values, not just the Coin.
+  -- By using 'areAllAdaOnly', the (True,True) case, this approach is both cheap and common,
+  -- and the computations of colInputsBal and totalBal, in the where clause, are not computed at all.
+  case (areAllAdaOnly utxoCollateral, areAllAdaOnly (txBody ^. collateralReturnTxBodyL)) of
+    (True, True) -> pure ()
+    (False, False) ->
+      if (Val.isAdaOnly totalBal)
+        then pure () -- Both have MultiSssets, and they cancel out.
+        else failureIf True $ CollateralContainsNonADA colInputBal
```

Comment:
I think we can use `failure $ CollateralContainsNonADA colInputBal` instead if we 
add `failure` to the imports above :
`import Validation (Validation, failure, failureIf, failureUnless,)`